### PR TITLE
Add backup related ffis.

### PR DESCRIPTION
**Pull request by overvenus**
Add backup related ffis.

cc pingcap/tikv#1290

PTAL @ngaut @siddontang @BusyJay

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: src/rocksdb.rs [hunk 1]
```diff
@@ -967,6 +967,78 @@ impl DBVector {
...
+        };
+
+        if let Err(e) = fs::create_dir_all(&Path::new(path)) {
+            return Err(format!("Failed to create rocksdb backup directory: \
+                                src/rocksdb.rs:                              \
+                                {:?}",
+                               e));
+        }
+
+        let be = unsafe { ffi_try!(rocksdb_backup_engine_open(opts.inner, cpath.as_ptr())) };
```
Comment (siddontang) [inline src/rocksdb.rs]: s/be/backup_engine/

**Comment by siddontang** [comment]
Rest LGTM

File: src/rocksdb.rs [hunk 2]
```diff
@@ -967,6 +967,79 @@ impl DBVector {
...
+                    .to_owned())
+            }
+        };
+
+        unsafe {
+            ffi_try!(rocksdb_backup_engine_restore_db_from_latest_backup(self.inner,
+                                                                         db_path.as_ptr(),
+                                                                         wal_path.as_ptr(),
+                                                                         ropts.inner))
+        };
```
Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: `rocksdb_backup_engine_restore_db_from_latest_backup` need 5 args

Comment (overvenus) [inline src/rocksdb.rs]: The missing `err` is hidden in `ffi_try!`.

Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: when and who free the err?

Comment (overvenus) [inline src/rocksdb.rs]: Also be freed in `ffi_try!`

```rust
macro_rules! ffi_try {
    ($func:ident($($arg:expr),*)) => ({
        use std::ptr;
        let mut err = ptr::null_mut();
        let res = $crate::$func($($arg),*, &mut err);
        if !err.is_null() {
            return Err($crate::error_message(err));
        }
        res
    })
}

pub fn error_message(ptr: *mut c_char) -> String {
    let c_str = unsafe { CStr::from_ptr(ptr) };
    let s = format!("{}", c_str.to_string_lossy());
    unsafe {
        libc::free(ptr as *mut c_void);
    }
    s
}
```

**Comment by zhangjinpeng1987** [comment]
LGTM

**Comment by overvenus** [comment]
PTAL @ngaut

**Comment by ngaut** [comment]
PTAL @BusyJay

File: src/rocksdb.rs [hunk 3]
```diff
@@ -967,6 +967,79 @@ impl DBVector {
...
+    pub fn new(opts: Options, path: &str) -> Result<BackupEngine, String> {
+        let cpath = match CString::new(path.as_bytes()) {
+            Ok(c) => c,
+            Err(_) => {
+                return Err("Failed to convert path to CString when opening rocksdb backup engine"
+                    .to_owned())
+            }
+        };
+
+        if let Err(e) = fs::create_dir_all(&Path::new(path)) {
```
Comment (BusyJay) [inline src/rocksdb.rs]: Just `path` should be ok.

File: src/rocksdb.rs [hunk 4]
```diff
@@ -967,6 +967,79 @@ impl DBVector {
...
+            return Err(format!("Failed to create rocksdb backup directory: \
+                                src/rocksdb.rs:                              \
```
Comment (BusyJay) [inline src/rocksdb.rs]: Why add the file name here?

File: src/rocksdb.rs [hunk 5]
```diff
@@ -967,6 +967,79 @@ impl DBVector {
...
+                               e));
+        }
+
+        let backup_engine =
+            unsafe { ffi_try!(rocksdb_backup_engine_open(opts.inner, cpath.as_ptr())) };
+
+        Ok(BackupEngine { inner: backup_engine })
+    }
+
+    pub fn create_new_backup(&self, db: &DB) -> Result<(), String> {
```
Comment (BusyJay) [inline src/rocksdb.rs]: I think moving this method into `DB` is better, for example `DB::backup_at(path, opt)`.

File: src/rocksdb.rs [hunk 6]
```diff
@@ -1113,6 +1186,36 @@ mod test {
...
+        let db_dir = TempDir::new("_rust_rocksdb_backuptest").unwrap();
+        let db = DB::open_default(db_dir.path().to_str().unwrap()).unwrap();
+        let p = db.put(key, value);
+        assert!(p.is_ok());
+
+        // Make a backup.
+        let backup_dir = TempDir::new("_rust_rocksdb_backuptest_backup").unwrap();
+        let backup_engine = BackupEngine::new(Options::new(), backup_dir.path().to_str().unwrap())
+            .unwrap();
+        let r = backup_engine.create_new_backup(&db);
```
Comment (BusyJay) [inline src/rocksdb.rs]: Just `unwrap`.

File: src/rocksdb_options.rs [hunk 7]
```diff
@@ -571,7 +565,7 @@ impl Options {
 
     pub fn set_ratelimiter(&mut self, rate_bytes_per_sec: i64) {
         let rate_limiter = RateLimiter::new(rate_bytes_per_sec,
-                                            100 * 1000 /* 100ms should work for most cases */,
+                                            100 * 1000, // 100ms should work for most cases
                                             10 /* should be good by leaving it at default 10 */);
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: Why this line is not changed?

Comment (BusyJay) [inline src/rocksdb_options.rs]: I think we can update this manually.

File: src/rocksdb_options.rs [hunk 8]
```diff
@@ -603,6 +597,31 @@ impl Drop for FlushOptions {
...
+pub struct RestoreOptions {
+    pub inner: *mut DBRestoreOptions,
+}
+
+impl RestoreOptions {
+    pub fn new() -> RestoreOptions {
+        unsafe { RestoreOptions { inner: rocksdb_ffi::rocksdb_restore_options_create() } }
+    }
+
+    pub fn set_keep_log_files(&self, flag: bool) {
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: Seems not covered.

File: src/rocksdb.rs (hunk 5 above)
Comment (BusyJay) [inline src/rocksdb.rs]: Why still keep these two methods?

File: src/rocksdb.rs [hunk 9]
```diff
@@ -806,6 +806,47 @@ impl DB {
     pub fn get_statistics(&self) -> Option<String> {
         self.opts.get_statistics()
     }
+
+    pub fn backup_at(&self, path: &str) -> Result<(), String> {
```
Comment (BusyJay) [inline src/rocksdb.rs]: Should return a `BackupEngine`.

File: src/rocksdb.rs (hunk 5 above)
Comment (BusyJay) [inline src/rocksdb.rs]: I don't see the connection between the redundant api and these abis.

Comment (overvenus) [inline src/rocksdb.rs]: Oops, my mistake. A better example:
```c++
// Ported
void rocksdb_backup_engine_create_new_backup(rocksdb_backup_engine_t* be,
                                             rocksdb_t* db, char** errptr) {
  SaveError(errptr, be->rep->CreateNewBackup(db->rep));
}

// Ported
void rocksdb_backup_engine_restore_db_from_latest_backup(
    rocksdb_backup_engine_t* be, const char* db_dir, const char* wal_dir,
    const rocksdb_restore_options_t* restore_options, char** errptr) {
  SaveError(errptr, be->rep->RestoreDBFromLatestBackup(std::string(db_dir),
                                                       std::string(wal_dir),
                                                       restore_options->rep));
}

// Not yet
void rocksdb_backup_engine_purge_old_backups(rocksdb_backup_engine_t* be,
                                             uint32_t num_backups_to_keep,
                                             char** errptr) {
  SaveError(errptr, be->rep->PurgeOldBackups(num_backups_to_keep));
}

// Not yet
const rocksdb_backup_engine_info_t* rocksdb_backup_engine_get_backup_info(
    rocksdb_backup_engine_t* be) {
  rocksdb_backup_engine_info_t* result = new rocksdb_backup_engine_info_t;
  be->rep->GetBackupInfo(&result->rep);
  return result;
}
```

File: src/rocksdb.rs [hunk 10]
```diff
@@ -967,6 +1006,40 @@ impl DBVector {
     }
 }
 
+pub struct BackupEngine {
+    inner: *mut DBBackupEngine,
+}
+
+impl BackupEngine {
+    pub fn new(opts: Options, path: &str) -> Result<BackupEngine, String> {
```
Comment (BusyJay) [inline src/rocksdb.rs]: `open` seems more appropriate.

File: src/rocksdb_options.rs (hunk 8 above)
Comment (BusyJay) [inline src/rocksdb_options.rs]: Any update?

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**